src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -347,21 +403,279 @@ ErrorIfModifyQueryNotSupported(Query *queryTree)
 	if (contain_mutable_functions((Node *) arbiterWhere) ||
 		contain_mutable_functions((Node *) onConflictWhere))
 	{
-		hasNonConstQualExprs = true;
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("functions used in the WHERE clause of the ON CONFLICT "
+							   "clause of INSERTs on distributed tables must be marked "
+								 "IMMUTABLE")));
 	}
 #endif
 
-	if (hasNonConstTargetEntryExprs || hasNonConstQualExprs)
+	if (specifiesPartitionValue)
 	{
 		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("functions used in modification queries on distributed "
-							   "tables must be marked IMMUTABLE")));
+						errmsg("modifying the partition value of rows is not allowed")));
 	}
```

Comment:
It looks likely that RETURNING support will be merged before this, so this will need to adapt.